[INDIV-197,INDIV-205] Add custom case update action for dfci swasth project

akashkj opened the pull request:
## Product Description
Added a custom automatic case update action for DFCI project. 
![image](https://github.com/dimagi/commcare-hq/assets/14318173/fe486cae-aa82-4880-816e-d10e3c614131)

## Technical Summary
https://dimagi-dev.atlassian.net/browse/INDIV-197
and
https://dimagi-dev.atlassian.net/browse/INDIV-205

Case property data is read from one case type `patient` and case property is updated in other case type `ccuser_caseload` that is related to it.

Case properties read - `screening_expiry_date` and `counselling_expiry_date`
Case properties updated - `counsellor_load` and `counsellor_closed_case_load`

## Feature Flag
NA

## Safety Assurance

### Safety story
A custom change, that require system admin role to use it in a project.
Tested using unit tests and QA testing on staging.

### Automated test coverage
Added unit tests

### QA Plan
https://dimagi-dev.atlassian.net/browse/QA-5535


### Migrations
NA

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

Commit:
Add custom action for dfci auto update rule + unit tests

akashkj commented:
@mkangia I have some testing pending, that will take some time to copy application and data between environments. Adding for review alongside to get the code reviewed.

Commit:
Refactor for missing values handling + related unit tests

mkangia reviewed (commented):
Hey @akashkj 
On a first pass this looks good. I think we should expect things to not be really ideal and handle all cases where things could be different, like expect case property missing or a wrong format of the case property value than expected.

akashkj commented:
> @mkangia I have some testing pending, that will take some time to copy application and data between environments. Adding for review alongside to get the code reviewed.

Changed https://github.com/dimagi/commcare-hq/pull/33364/commits/a2c203ab163b886614d598d75bb6c4eacd2e4170

mkangia commented on custom/dfci_swasth/rules/custom_actions.py:
```diff
@@ -0,0 +1,84 @@
+    coun_ccuser_caseload_case_id = patient_case.get_case_property(PROP_CCUSER_CASELOAD_CASE_ID)
+    try:
+        return CommCareCase.objects.get_case(coun_ccuser_caseload_case_id, domain=patient_case.domain)
+    except (CaseNotFound, CommCareCase.DoesNotExist):
```
I think its a bit confusing to have two exceptions here.
I feel this would be cleaner as
```
coun_ccuser_caseload_case_id = patient_case.get_case_property(PROP_CCUSER_CASELOAD_CASE_ID)
if coun_ccuser_caseload_case_id:
    try:
        return CommCareCase.objects.get_case(coun_ccuser_caseload_case_id, domain=patient_case.domain)
    except CommCareCase.DoesNotExist:
        return None
```

akashkj replied:
Changed

mkangia commented on custom/dfci_swasth/rules/custom_actions.py:
```diff
@@ -0,0 +1,84 @@
+    try:
+        return datetime.strptime(screening_expiry_date, ISO_DATE_FORMAT)
+    except ValueError:
+        return None
```
Similar thing can be done here as well
```
screening_expiry_date = patient_case.get_case_property(property_name)
if screening_expiry_date:
    try:
        return datetime.strptime(screening_expiry_date, ISO_DATE_FORMAT)
    except ValueError:
        return None
```

mkangia commented on custom/dfci_swasth/rules/custom_actions.py:
```diff
@@ -0,0 +1,84 @@
+
+
+def _get_updated_counsellor_load(ccuser_caseload_case):
+    counsellor_load = int(ccuser_caseload_case.get_case_property(PROP_COUNSELLOR_LOAD))
```
What would happen if this case property is missing? or a string that is not a number?

akashkj replied:
This will throw `TypeError` and `ValueError` respectively which are caught in the calling function.

mkangia replied:
Cool, so what happens then? are those exceptions handled gracefully in the workflow somewhere?

mkangia commented on custom/dfci_swasth/rules/custom_actions.py:
```diff
@@ -0,0 +1,84 @@
+    try:
+        return {PROP_COUNSELLOR_LOAD: _get_updated_counsellor_load(ccuser_caseload_case)}
+    except (TypeError, ValueError):
+        return {}
```
We should also avoid updating the case if there are no updates to be done.

mkangia replied:
And this should also be logged as an error so its possible to identify why this failed.

akashkj replied:
> We should also avoid updating the case if there are no updates to be done.

Do you mean to return something like `CaseRuleActionResult(num_related_updates=0)` in this case?

mkangia replied:
No I meant we should avoid the update_case call. Looks like we would do an empty update on the case

akashkj replied:
Added check for non-empty update

mkangia commented on custom/dfci_swasth/rules/custom_actions.py:
```diff
@@ -0,0 +1,84 @@
+            device_id=__name__ + ".update_counsellor_load",
+            form_name=rule.name,
+        )
+        num_updates = 1
```
And should this be num_updates or num_related_updates since we are updating a related case?

akashkj replied:
Changed to `num_related_updates`

mkangia commented on custom/dfci_swasth/tests/test_custom_actions.py:
```diff
@@ -0,0 +1,153 @@
+        super().tearDown()
+
+    def test_non_patient_case(self):
+        patient_rule = create_empty_rule(
```
Any reason to create a new one for each test?

mkangia replied:
creating one in setup would also make it easy to delete it in teardown.

akashkj replied:
Updated

mkangia commented on custom/dfci_swasth/tests/test_custom_actions.py:
```diff
@@ -0,0 +1,153 @@
+        )
+
+        result = update_counsellor_load(patient_case, patient_rule)
+        self.assertEqual(0, result.num_updates)
```
this assertion is really not specific here. num_updates could be 0 for different reasons, hence my concern with the implementation of the function update_counsellor_load itself.
If there was specific response that would correspond to unexpected case type, that would be the ideal test here.

akashkj replied:
> specific response

Any suggestion how can we achieve that? 
I have referred existing custom action from https://github.com/dimagi/commcare-hq/blob/bf8245fd2e7ac461ead1ae54737bd8dcec1b3e1d/custom/gcc_sangath/rules/custom_actions.py#L25 to handle no update scenario.

mkangia replied:
can we raise an exception?

akashkj replied:
If we raise an exception from `update_counsellor_load()`, calling action executor should be able to handle it.

akashkj commented on custom/dfci_swasth/rules/custom_actions.py:
```diff
@@ -0,0 +1,84 @@
+    num_updates = 0
+
+    if patient_case.type != CASE_TYPE_PATIENT:
+        return CaseRuleActionResult(num_updates=num_updates)
```
There is `num_errors`, will that provide the info?

mkangia replied:
I think we need a way to store some error message as well

akashkj replied:
I couldn't find any argument used in `CaseRuleActionResult` to store the error message, it just deals with different type of counts.

Commit:
Update test_custom_actions.py

Commit:
Remove expiry date checks from the custom action + tests update

akashkj commented:
@mkangia Updated the code to exclude expiry date related checks. Also updated the relevant tests.

mkangia commented:
Hey @akashkj 

Let me know when this is ready for review.
I see there is some pending feedback.

akashkj commented:
> Hey @akashkj
> 
> Let me know when this is ready for review. I see there is some pending feedback.

@mkangia Addressed the comments. It can be reviewed now.

mkangia commented on custom/dfci_swasth/rules/custom_actions.py:
```diff
@@ -0,0 +1,66 @@
+        # Value of counsellor load can be minimum 0
+        if counsellor_load >= 1:
+            return counsellor_load - 1
+    except (TypeError, ValueError):
```
Can you share what scenarios these 2 exceptions are handling each?
we already checked for None earlier.

Additionally, good to wrap the code that is expected to raise the exception itself, so it's clear the code that is expected to raise.
```
counsellor_load_raw = ccuser_caseload_case.get_case_property(PROP_COUNSELLOR_LOAD)
if counsellor_load_raw:
    try:
        counsellor_load = int(counsellor_load_raw)
    except (TypeError, ValueError):
        pass
    else:
        if counsellor_load >= 1:
            return counsellor_load - 1
```

akashkj replied:
`ValueError` will check for non-numeric value and `TypeError` will check for typecasting with `None` value.
Modified the code to raise exception from place where it is originating.
```
def _get_integer_case_property_value(case, property_name):
    try:
        return int(case.get_case_property(property_name))
    except (TypeError, ValueError):
        return
```

mkangia commented on custom/dfci_swasth/rules/custom_actions.py:
```diff
@@ -0,0 +1,66 @@
+    counsellor_load = _get_updated_counsellor_load(ccuser_caseload_case)
+    if not counsellor_load:
+        return {}
+    return {PROP_COUNSELLOR_LOAD: counsellor_load}
```
I guess you can simplify this to
```
counsellor_load = _get_updated_counsellor_load(ccuser_caseload_case)
if counsellor_load:
    return {PROP_COUNSELLOR_LOAD: counsellor_load}
```

akashkj replied:
Added changes for https://dimagi-dev.atlassian.net/browse/INDIV-205 as well and modified the code accordingly.
```
def _get_case_updates(ccuser_caseload_case):
    counsellor_load = _get_updated_counsellor_load(ccuser_caseload_case)
    counsellor_closed_case_load = _get_updated_counsellor_closed_case_load(ccuser_caseload_case)

    result = {}
    if counsellor_load is not None:
        result.update({PROP_COUNSELLOR_LOAD: counsellor_load})
    if counsellor_closed_case_load is not None:
        result.update({PROP_COUNSELLOR_CLOSED_CASE_LOAD: counsellor_closed_case_load})
    return result
```

Commit:
INDIV-205 increment 'counsellor closed case load' by 1

sravfeyn commented on custom/dfci_swasth/rules/custom_actions.py:
```diff
@@ -0,0 +1,81 @@
+
+
+def _get_ccuser_caseload_case(patient_case):
+    coun_ccuser_caseload_case_id = patient_case.get_case_property(PROP_CCUSER_CASELOAD_CASE_ID)
```
nit: Could rename `coun_ccuser_caseload_case_id` to simply `case_id` since there is just one `case_id` var in this function

akashkj replied:
Changed
https://github.com/dimagi/commcare-hq/compare/2fb18c32d0b1fbf2d7e7d8c79a5e43753f7f7433..b4454ac55a76b9d95340aae30b0561cae6bdbf6a

sravfeyn commented on custom/dfci_swasth/rules/custom_actions.py:
```diff
@@ -0,0 +1,81 @@
+    counsellor_closed_case_load = _get_integer_case_property_value(
+        ccuser_caseload_case, PROP_COUNSELLOR_CLOSED_CASE_LOAD)
+    # Value of counsellor closed case load can be minimum 0
+    if counsellor_closed_case_load is not None and counsellor_closed_case_load >= 0:
```
```suggestion
    if counsellor_closed_case_load and counsellor_closed_case_load >= 0:
```

akashkj replied:
Same as https://github.com/dimagi/commcare-hq/pull/33364#discussion_r1314645669

sravfeyn replied:
👍  Got it.

mkangia replied:
It's strange though that for counsellor_load the check is not for None, I would assume that the reasoning would apply to both?

akashkj replied:
For `counsellor_load`, there is check for it to be one or more, so if its zero, whole condition will still be false . While in case of `counsellor_closed_case_load` it can be zero or more, hence the `None` check as value zero will otherwise evaluate to false.

mkangia replied:
the comment for both is "can be minimum 0", so that needs to change for one of them?

akashkj replied:
Updated comment
https://github.com/dimagi/commcare-hq/compare/2812e5ac55d27a3b34da11e3fdff6122083603dc..48ecc72c9e560e9c7f2e02192f534f2367f31c08

sravfeyn commented on custom/dfci_swasth/rules/custom_actions.py:
```diff
@@ -0,0 +1,81 @@
+    num_related_updates = 0
+
+    if patient_case.type != CASE_TYPE_PATIENT:
+        return CaseRuleActionResult(num_related_updates=num_related_updates)
```
Since  the `num_related_updates` kwarg is zero by default
```suggestion
        return CaseRuleActionResult()
```

akashkj commented on custom/dfci_swasth/rules/custom_actions.py:
```diff
@@ -0,0 +1,81 @@
+def _get_updated_counsellor_load(ccuser_caseload_case):
+    counsellor_load = _get_integer_case_property_value(ccuser_caseload_case, PROP_COUNSELLOR_LOAD)
+    # Value of counsellor load can be minimum 0
+    if counsellor_load is not None and counsellor_load >= 1:
```
Added explicit `None` check here as value of `counsellor_load` can be zero as well. In that case, `if counsellor_load` check will evaluate to false which is wrong for this case.

sravfeyn replied:
The [below line](https://github.com/dimagi/commcare-hq/pull/33364/files#r1314645906) is okay since you are 'trueing' on zero (`>=0`) on 2nd condition, but here you are 'trueing' only on `>= 1`, which means this condition could simply be `if counsellor_load`.

akashkj commented:
Updated PR code and description to include https://dimagi-dev.atlassian.net/browse/INDIV-205 changes as well.

Commit:
Update conditional check

Co-authored-by: Sravan Reddy <sravfeyn@gmail.com>

sravfeyn commented on custom/dfci_swasth/rules/custom_actions.py:
```diff
@@ -0,0 +1,81 @@
+
+def _get_updated_counsellor_load(ccuser_caseload_case):
+    counsellor_load = _get_integer_case_property_value(ccuser_caseload_case, PROP_COUNSELLOR_LOAD)
+    # Value of counsellor load can be minimum 0
+    if counsellor_load and counsellor_load >= 1:
```
Note that the comment says minimum 0 and the condition says minimum 1. If it indeed is minimum 1, this can be updated as below
```suggestion
    # Value of counsellor load should be 1 or more
    if counsellor_load:
```

akashkj replied:
Value of `counsellor_load` can be minimum zero and in this code we decrement its value by 1. Due to this, have put check that it should be minimum 1 so that on decrementing it become minimum possible value 0. Maybe I can update the same in the comment as well.

sravfeyn reviewed (approved):
Thanks for all prompt followups!

Commit:
Merge remote-tracking branch 'origin/master' into akj/indiv-197-dfci-custom-update-rule

Commit:
Update comment

Commit:
Update comment
